src/main/java/com/amazon/opendistroforelasticsearch/sql/rewriter/join/JoinRewriteRule.java

```diff
@@ -0,0 +1,191 @@
+ *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ *   express or implied. See the License for the specific language governing
+ *   permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.sql.rewriter.join;
+
+import com.alibaba.druid.sql.ast.expr.SQLIdentifierExpr;
+import com.alibaba.druid.sql.ast.expr.SQLQueryExpr;
+import com.alibaba.druid.sql.ast.statement.SQLExprTableSource;
+import com.alibaba.druid.sql.ast.statement.SQLJoinTableSource;
+import com.alibaba.druid.sql.ast.statement.SQLSelectQuery;
+import com.alibaba.druid.sql.dialect.mysql.ast.statement.MySqlSelectQueryBlock;
+import com.alibaba.druid.sql.dialect.mysql.visitor.MySqlASTVisitorAdapter;
+import com.amazon.opendistroforelasticsearch.sql.esdomain.LocalClusterState;
+import com.amazon.opendistroforelasticsearch.sql.esdomain.mapping.FieldMappings;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.RewriteRule;
+import com.amazon.opendistroforelasticsearch.sql.rewriter.matchtoterm.VerificationException;
+import com.amazon.opendistroforelasticsearch.sql.utils.StringUtils;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.Multimap;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Consumer;
+
+/**
+ * Rewrite rule for query involved JOIN.
+ */
+public class JoinRewriteRule implements RewriteRule<SQLQueryExpr> {
+
+    private static final String DOT = ".";
+    private int aliasSuffix = 0;
+    private final LocalClusterState clusterState;
+
+    public JoinRewriteRule(LocalClusterState clusterState) {
+        this.clusterState = clusterState;
+    }
+
+    @Override
+    public boolean match(SQLQueryExpr root) {
+        return isJoin(root);
+    }
+
+    private boolean isJoin(SQLQueryExpr sqlExpr) {
+        SQLSelectQuery sqlSelectQuery = sqlExpr.getSubQuery().getQuery();
+
+        if (!(sqlSelectQuery instanceof MySqlSelectQueryBlock)) {
+            return false;
+        }
+
+        MySqlSelectQueryBlock query = (MySqlSelectQueryBlock) sqlSelectQuery;
+        return query.getFrom() instanceof SQLJoinTableSource
+            && ((SQLJoinTableSource) query.getFrom()).getJoinType() != SQLJoinTableSource.JoinType.COMMA;
+    }
+
+    @Override
+    public void rewrite(SQLQueryExpr root) {
+
+        final Multimap<String, Table> tableByFieldName = ArrayListMultimap.create();
+        final Set<String> tableNameAndAlias = new HashSet<>();
+        final Map<String, String> tableNameToAlias = new HashMap<>();
+
+        visitTable(root, tableExpr -> {
+            // Copied from SubqueryAliasRewriter ; Removes index type name if any
+            String tableName = tableExpr.getExpr().toString().replaceAll(" ", "").split("/")[0];
+
+            if (tableExpr.getAlias() == null) {
+                // Could we not directly use table name as alias?
```

Comment:
If the table join with itself, the alias will be duplicate. not sure wheter the syntax or semantic check this before reach here.

Reply:
For self join, the alias will not be duplicate as we generate different alias for each table without alias. I have added more checks to handle joins with same tables on either side. Added more integration tests as well.